Use new XMLString type for CAPA data field.

jcdyer opened the pull request:
### [TNL-5245](https://openedx.atlassian.net/browse/TNL-5245)

### Reviewers

- [ ] @nasthagiri 
- [ ] One volunteer from @edx/platform-team

### Post-review

- [ ] Squash commits.

doctoryes commented on requirements/edx/github.txt:
```diff
@@ -71,7 +71,7 @@ git+https://github.com/edx/rfc6266.git@v0.0.5-edx#egg=rfc6266==0.0.5-edx
 
 # Our libraries:
-git+https://github.com/edx/XBlock.git@xblock-0.4.12#egg=XBlock==0.4.12
+git+https://github.com/edx/XBlock.git@cdyer/xml-string-field#egg=XBlock==0.5.0a1
```
Reminder: change this!

jcdyer replied:
Will do.

nasthagiri commented on common/lib/xmodule/xmodule/tests/test_capa_module.py:
```diff
@@ -2487,18 +2481,15 @@ def test_indexing_problem_with_html_tags(self):
     def test_invalid_xml_handling(self):
         """
-        Tests to confirm that invalid XML does not throw a wake-up-ops level error.
-        See TNL-5057 for quick fix, TNL-5245 for full resolution.
```
Yay to removal of tech-debt.

nasthagiri commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -76,7 +76,7 @@ class ComplexEncoder(json.JSONEncoder):
     Extend the JSON encoder to correctly handle complex numbers
     """
-    def default(self, obj):
+    def default(self, obj):  # pylint: disable=method-hidden
```
What is the underlying cause of this pylint issue?  Is it not a real issue?

jcdyer replied:
It's caused by the design of the json library.  `json.encode` lets you pass a `default()` function that overrides the builtin `default` method.

nasthagiri replied:
got it.  thanks.

jcdyer commented:
Fixed some code that I broke in refactoring, and resolved quality issue.

jcdyer commented:
jenkins run all

jcdyer commented:
Some legitimate bokchoy errors.  I'll work on resolving those.

jcdyer commented:
jenkins run bokchoy

jcdyer commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -157,7 +158,12 @@ class CapaFields(object):
+    data = XMLString(
+        help=_("XML data for the problem"),
+        scope=Scope.content,
+        enforce_type=FEATURES.get('ENABLE_XBLOCK_XML_VALIDATION', True),
```
This has (yuck) a dependency on the django settings.  This module was already using settings, but I'd love suggestions for how to avoid that...

cpennington replied:
Sadly, we don't have a current standard for a way to set field-level values at runtime through xblock-specific apis.

jcdyer commented:
@nasthagiri @doctoryes I've changed a few things on this PR.  To address @cpennington's concerns about too much xml parsing, I moved the validation from deserialization to serialization, so it should only perform the extra parse on write, not on read.  Another solution would be to create a parsed xml field where the deserialized value is the lxml etree object, but that would require diving into the guts of capa, and reorganizing a number of things.

cpennington commented:
This PR is being deployed to edx.org Staging in preparation for a release to production on Tuesday, 12/20 at 10 am. If you feel you need to validate it manually, please do so by then, and contact me if there is an urgent enough issue to scrap the deployment.

cpennington commented:
This PR has been deployed to edx.org Staging in preparation for a release to production on Wednesday, 12/21/16 at 10 am. If you feel you need to validate it manually, please do so by then, and contact me if there is an urgent enough issue to scrap the deployment.
